lib/Sema/CSBindings.cpp

```diff
@@ -75,13 +83,284 @@ void ConstraintSystem::inferTransitiveSupertypeBindings(
+      bool isDirectRequirement = entry.second.first;
+      bool &isCovered = entry.second.second;
+
+      if (isCovered)
+        continue;
+
+      // FIXME: This is a hack and it's incorrect because it depends
+      // on ordering of the literal procotols e.g. if `ExpressibleByNilLiteral`
+      // appears before e.g. `ExpressibleByIntegerLiteral` we'd drop
+      // optionality although that would be incorrect.
+      do {
+        // If the type conforms to this protocol, we're covered.
+        if (TypeChecker::conformsToProtocol(type, protocol, cs.DC)) {
+          isCovered = true;
+          break;
+        }
+
+        // If this literal protocol is not a direct requirement it
+        // would be possible to change optionality while inferring
+        // bindings for a supertype, so this hack doesn't apply.
+        if (!isDirectRequirement)
+          break;
+
+        // If we're allowed to bind to subtypes, look through optionals.
+        // FIXME: This is really crappy special case of computing a reasonable
+        // result based on the given constraints.
+        if (binding.Kind == AllowedBindingKind::Subtypes) {
+          if (auto objTy = type->getOptionalObjectType()) {
+            requiresUnwrap = true;
+            type = objTy;
+            continue;
+          }
+        }
+
+        requiresUnwrap = false;
+        break;
+      } while (true);
+    }
+
+    if (requiresUnwrap)
+      binding.BindingType = type;
+  }
+
+  // If this is not a literal protocol or it has been "covered" by an existing
+  // binding it can't provide a default type.
+  auto isUnviableForDefaulting = [&literalProtocols](ProtocolDecl *protocol) {
+    auto literal = literalProtocols.find(protocol);
+    return literal == literalProtocols.end() || literal->second.second;
+  };
+
+  for (auto *constraint : Protocols) {
+    auto *protocol = constraint->getProtocol();
+
+    if (isUnviableForDefaulting(protocol))
+      continue;
+
+    // Let's try to coalesce integer and floating point literal protocols
+    // if they appear together because the only possible default type that
+    // could satisfy both requirements is `Double`.
+    if (protocol->isSpecificProtocol(
+          KnownProtocolKind::ExpressibleByIntegerLiteral)) {
+      auto *floatLiteral = cs.getASTContext().getProtocol(
+          KnownProtocolKind::ExpressibleByFloatLiteral);
+      // If `ExpressibleByFloatLiteral` is a requirement and it isn't
+      // covered, let's skip `ExpressibleByIntegerLiteral` requirement.
+      if (!isUnviableForDefaulting(floatLiteral))
+        continue;
+    }
+
+    auto defaultType = TypeChecker::getDefaultType(protocol, cs.DC);
+    if (!defaultType)
+      continue;
+
+    if (isUnviableDefaultType(defaultType, existingTypes))
+      continue;
+
+    // We need to figure out whether this is a direct conformance
+    // requirement or inferred transitive one to identify binding
+    // kind correctly.
```

Comment:
I don't fully understand this - could you elaborate a little on why this is the case?

Reply:
Sure! If the conformance requirement is inferred transitively solver should be allowed to try supertypes of default type, if it’s directly associated with the current type variable solver is allowed to try subtypes of default. Does that make more sense?

Reply:
So if we have `$T0 <: $T1` and `$T0 conforms to ExpressibleByIntegerLiteral` for example, when inferring a default for `$T1` from the conformance requirement on `$T0` solver should be allowed to attempt supertypes just like if it was `Int <: T1`.

Reply:
Ohhh okay, I was thinking about transitive bindings the other way around (e.g. the conformance is on the supertype instead of the subtype). This makes sense. I realize now that the documentation for `inferTransitiveBindings ` would have cleared up my confusion haha. Thank you!